### PR TITLE
Add support for other bucket aggregates in histogram

### PR DESCRIPTION
**Pull request by RobAtticus**
The histogram aggregate previously only offered support for doing
a count in each bucket. This PR adds support for other functions:
min, max, sum, and avg.

Addresses https://github.com/timescale/timescaledb/issues/272

### CODE REVIEW

**Comment by PaulGrimshaw** [comment]
This looks great. Don't have enough knowledge of the library to meaningfully comment on the implementation, but believe this is a very useful feature.

File: src/histogram.c [hunk 1]
```diff
@@ -5,16 +5,43 @@
...
+{
+	SUBAGG_AVG = 0,
+	SUBAGG_COUNT,
+	SUBAGG_MIN,
+	SUBAGG_MAX,
+	SUBAGG_SUM,
+	_MAX_SUBAGG,
+} SubAgg;
+
+static const char *subagg_names[_MAX_SUBAGG + 1] = {
```
Comment (erimatnor) [inline src/histogram.c]: Rather than hard-coding the supported aggregate functions, can't we do this for the general case. Internally, I believe we should be able to check whether a given function `OID` is an aggregate function with the right input and output. Look at, e.g., `parse_agg.{c,h}`, `pg_aggregate.c` and all the `get_func_` functions in `lsyscache.c` for insight. For instance, `lookup_agg_function()` sounds useful (although it is static).

File: src/histogram.c [hunk 2]
```diff
@@ -57,32 +111,78 @@ hist_sfunc(PG_FUNCTION_ARGS)
...
-	hist = (Datum *) VARDATA(state);
-	hist[bucket] = Int32GetDatum(DatumGetInt32(hist[bucket]) + 1);
+	/* Update histogram based on bucket_agg/subagg */
+	hist = (Datum *) VARDATA(state->values);
+	hist_opt = (Datum *) VARDATA(state->opt_values);
 
+	hist_val = DatumGetFloat8(hist[bucket]);
+	hist_opt_val = DatumGetFloat8(hist_opt[bucket]);
+	val = DatumGetFloat8(val_datum);
+	switch (state->subagg)
```
Comment (erimatnor) [inline src/histogram.c]: See my comment above about trying to do this for the general case.

File: src/histogram.c [hunk 3]
```diff
@@ -135,65 +295,86 @@ hist_combinefunc(PG_FUNCTION_ARGS)
...
+	pq_begintypsend(&buf);
+	pq_sendint(&buf, VARSIZE(state->values) - VARHDRSZ, 4);
+	pq_sendbytes(&buf, VARDATA(state->values), VARSIZE(state->values) - VARHDRSZ);
 
-	for (i = 0; i < HIST_LEN(state); i++)
-	{
-		hist[i] = Int32GetDatum(htonl(DatumGetInt32(hist[i])));
-	}
+	pq_sendint(&buf, VARSIZE(state->opt_values) - VARHDRSZ, 4);
+	pq_sendbytes(&buf, VARDATA(state->opt_values), VARSIZE(state->opt_values) - VARHDRSZ);
```
Comment (erimatnor) [inline src/histogram.c]: When using these serialization functions, I think we can get rid of the winsock lib dependency for the windows build that was needed due to `htonl`.

File: sql/histogram.sql [hunk 4]
```diff
@@ -1,8 +1,14 @@
-CREATE OR REPLACE FUNCTION _timescaledb_internal.hist_sfunc (state INTERNAL, val DOUBLE PRECISION, MIN DOUBLE PRECISION, MAX DOUBLE PRECISION, nbuckets INTEGER)
+CREATE OR REPLACE FUNCTION _timescaledb_internal.hist_sfunc (state INTERNAL, val DOUBLE PRECISION, MIN DOUBLE PRECISION, MAX DOUBLE PRECISION, nbuckets INTEGER, bucket_agg TEXT = 'count')
```
[the remark this answers is not in the document]
Comment (RobAtticus) [reply sql/histogram.sql]: (Just a reminder from our convo that this won't work because of overloaded functions)

**Comment by RobAtticus** [comment]
@erimatnor @cevian this is ready for review

File: sql/histogram.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply sql/histogram.sql]: Got it.

File: src/histogram.c [hunk 5]
```diff
@@ -29,12 +38,226 @@ TS_FUNCTION_INFO_V1(hist_serializefunc);
...
+	Datum		res;
+
+	fmgr_info(funcid, &flinfo);
+	InitFunctionCallInfoData(fcinfo, &flinfo, 2, InvalidOid, ctxt, NULL);
+	fcinfo.arg[0] = arg1;
+	fcinfo.arg[1] = arg2;
+	fcinfo.argnull[0] = false;
+	fcinfo.argnull[1] = false;
+
+	res = FunctionCallInvoke(&fcinfo);
```
Comment (erimatnor) [inline src/histogram.c]: Any reason we are not using `OidFunctionCall2()` here? I think that would be a lot cleaner if possible. Is it because you need to pass on the context via `ctxt`?

File: src/histogram.c [hunk 6]
```diff
@@ -29,12 +38,226 @@ TS_FUNCTION_INFO_V1(hist_serializefunc);
...
```
Comment (erimatnor) [inline src/histogram.c]: Same here: use `OidFunctionCall2()` if possible.

File: src/histogram.c [hunk 7]
```diff
@@ -29,12 +38,226 @@ TS_FUNCTION_INFO_V1(hist_serializefunc);
...
+	FmgrInfo	flinfo;
+	FunctionCallInfoData fcinfo;
+	Datum		res;
+
+	fmgr_info(funcid, &flinfo);
+	InitFunctionCallInfoData(fcinfo, &flinfo, 2, InvalidOid, ctxt, NULL);
+	fcinfo.arg[0] = arg1;
+	fcinfo.argnull[0] = false;
+
+	res = FunctionCallInvoke(&fcinfo);
```
Comment (erimatnor) [inline src/histogram.c]: Possibly use `OidFunctionCall1()`

File: src/histogram.c [hunk 8]
```diff
@@ -44,45 +267,119 @@ hist_sfunc(PG_FUNCTION_ARGS)
...
-	{
-		/* cannot be called directly because of internal-type argument */
 		elog(ERROR, "hist_sfunc called in non-aggregate context");
-	}
 
+	/* Cannot generate a histogram with incompatible bounds */
 	if (min > max)
-	{
-		/* cannot generate a histogram with incompatible bounds */
 		elog(ERROR, "lower bound cannot exceed upper bound");
```
Comment (erimatnor) [inline src/histogram.c]: We should try to use `ereport()` with proper error code and optional details/hints when generating user-induced errors (e.g., user passed illegal arguments, etc.)

File: src/histogram.c [hunk 9]
```diff
@@ -44,45 +267,119 @@ hist_sfunc(PG_FUNCTION_ARGS)
...
-	}
+
+	/* Check if bucket_agg is set, otherwise use 'count' */
+	if (PG_NARGS() == 5)
```
Comment (erimatnor) [inline src/histogram.c]: Isn't it better to use `PG_ARGISNULL(5)`. Seems more explicit since the "other" case could be that `PG_NARGS()` is anything but 5.

File: src/histogram.c [hunk 10]
```diff
@@ -29,12 +38,226 @@ TS_FUNCTION_INFO_V1(hist_serializefunc);
 TS_FUNCTION_INFO_V1(hist_deserializefunc);
 TS_FUNCTION_INFO_V1(hist_finalfunc);
 
-/* histogram(state, val, min, max, nbuckets) */
+typedef struct SubAggState
```
Comment (erimatnor) [inline src/histogram.c]: Can we document the roles of each of these structs? It is a bit hard to follow right now what each of them are used for.

File: src/histogram.c [hunk 11]
```diff
@@ -44,45 +267,119 @@ hist_sfunc(PG_FUNCTION_ARGS)
...
+		subagg_text = TextDatumGetCString(PG_GETARG_CSTRING(5));
 
 	if (state == NULL)
 	{
-		/* Allocate memory to a new histogram state array */
-		Size		arrsize = sizeof(Datum) * (nbuckets + 2);
+		/* Length is nbuckets plus 1 on either side */
+		Size		arrsize = nbuckets + 2;
+		HistogramInternalState *istate =
+		(HistogramInternalState *) fcinfo->flinfo->fn_extra;
```
Comment (erimatnor) [inline src/histogram.c]: I don't recall why this isn't part of or tagged on to the `HistogramState`?

File: src/histogram.c [hunk 12]
```diff
@@ -92,108 +389,208 @@ Datum
...
+	pq_sendint(&buf, state->subagg_state.combinefunc, sizeof(Oid));
+	pq_sendint(&buf, state->subagg_state.serialfunc, sizeof(Oid));
+	pq_sendint(&buf, state->subagg_state.deserialfunc, sizeof(Oid));
+
+	/* no serialfn given, get info to serialize each bucket based on type */
+	if (!OidIsValid(state->subagg_state.serialfunc) && !OidIsValid(sstate->fn))
+	{
+		bool		is_varlen;
+
+		getTypeBinaryOutputInfo(state->subagg_state.result_type, &sstate->fn,
```
Comment (erimatnor) [inline src/histogram.c]: Should we check the result of `is_varlen`? Does it matter?

File: src/histogram.c (hunk 11 above)
Comment (erimatnor) [reply src/histogram.c]: Also, if `HistogramInternalState` is in `fn_extra`, perhaps a better name is `HistogramInternalFunctionState` or something to indicate it is basically a function call cache. Not a biggie though.

File: src/histogram.c (hunk 5 above)
Comment (RobAtticus) [reply src/histogram.c]: Need to pass `ctxt` so unfortunately cannot (same below)

File: src/histogram.c (hunk 12 above)
Comment (RobAtticus) [reply src/histogram.c]: I did steal this from bookends code but I don't believe it matters.

File: src/histogram.c (hunk 11 above)
Comment (RobAtticus) [reply src/histogram.c]: Done

File: src/histogram.c (hunk 9 above)
Comment (RobAtticus) [reply src/histogram.c]: Done

File: src/histogram.c [hunk 13]
```diff
@@ -29,12 +38,226 @@ TS_FUNCTION_INFO_V1(hist_serializefunc);
...
+	 * if we can find a NUMERIC or ANY version of the aggregator
+	 */
+	while (funclist != NULL && (funclist->nargs != 1 ||
+								(funclist->args[0] != FLOAT8OID && funclist->args[0] != ANYOID)))
+	{
+		funclist = funclist->next;
+	}
+
+	if (NULL == funclist)
+		elog(ERROR, "Aggregation function not found");
```
[the remark this answers is not in the document]
Comment (RobAtticus) [reply src/histogram.c]: Done

File: src/histogram.c (hunk 10 above)
Comment (RobAtticus) [reply src/histogram.c]: Done

File: src/histogram.c (hunk 9 above)
Comment (RobAtticus) [reply src/histogram.c]: Actually I tried this and it kept crashing. I am not sure why.

File: src/errors.h [hunk 14]
```diff
@@ -10,6 +10,7 @@
 #define ERRCODE_IO_QUERY_ERRORS MAKE_SQLSTATE('I','O','0','0','0')
 #define ERRCODE_IO_HYPERTABLE_NOT_EXIST MAKE_SQLSTATE('I','O','0','0','1')
 #define ERRCODE_IO_DIMENSION_NOT_EXIST MAKE_SQLSTATE('I','O','0','0','2')
+#define ERRCODE_IO_HISTOGRAM_MAX_LESS_MIN MAKE_SQLSTATE('I','O','0','0','3')
```
Comment (erimatnor) [inline src/errors.h]: Do we need a new error code for this? Seem extremely specific. Won't a `ERRCODE_INVALID_PARAMETER_VALUE` suffice?

File: src/histogram.c (hunk 12 above)
Comment (erimatnor) [reply src/histogram.c]: Can we verify that it doesn't matter?

Comment (RobAtticus) [reply src/histogram.c]: It doesn't matter since it is used to tell whether the return type from the serialization function is fixed or variable length, but we get the size of the returned bytes later using the `VARSIZE` macro. So whether it is fixed or variable doesn't matter for our purposes, we get the size anyway.

File: src/histogram.c [hunk 15]
```diff
@@ -29,12 +39,240 @@ TS_FUNCTION_INFO_V1(hist_serializefunc);
...
+	HistogramState *state = (HistogramState *) MemoryContextAlloc(aggcontext, sizeof(HistogramState));
+
+	state->values = MemoryContextAllocZero(aggcontext, VALUES_SIZE(arrsize));
+	state->first_tuple = MemoryContextAlloc(aggcontext, arrsize);
+	SET_VARSIZE(state->values, VALUES_SIZE(arrsize));
+	set_first_tuples(state, true);
+	state->subagg_state = (SubAggState)
+	{
+		.sfunc = InvalidOid,
+			.combinefunc = InvalidOid,
```
[the remark this answers is not in the document]
Comment (RobAtticus) [reply src/histogram.c]: pgindent refuses to change it.

File: src/histogram.c [hunk 16]
```diff
@@ -92,108 +406,208 @@ Datum
 hist_combinefunc(PG_FUNCTION_ARGS)
```
[the remark this answers is not in the document]
Comment (RobAtticus) [reply src/histogram.c]: check added

File: src/histogram.c [hunk 17]
```diff
@@ -29,12 +39,240 @@ TS_FUNCTION_INFO_V1(hist_serializefunc);
...
+			hist[i] = get_agg_init_val(istate->text_init_value,
+									   istate->trans_type, ctxt);
+		set_first_tuples(state, false);
+	}
+}
+
+static Datum
+apply_subagg_sfunc(Oid funcid, fmNodePtr ctxt, Datum arg1, Datum arg2)
+{
+	FmgrInfo	flinfo;
```
[the remark this answers is not in the document]
Comment (RobAtticus) [reply src/histogram.c]: will check

File: src/histogram.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/histogram.c]: Maybe just make this non-C99. Anyhow seems a bit weird to me to initialize a heap object this way.

File: sql/histogram.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ngalstyan4) [reply sql/histogram.sql]: @RobAtticus should it not at least be `NAME` and not `TEXT`  ?

**Comment by RobAtticus** [comment]
Closing because I don't think we're going with this approach after all.
